zookeeper-server/src/main/java/org/apache/zookeeper/server/auth/X509AuthenticationConfig.java

```diff
@@ -0,0 +1,324 @@
+  private static X509AuthenticationConfig instance = null;
+  private static final String LOG_MSG_PREFIX = X509AuthenticationConfig.class.getName() + "::";
+
+  private X509AuthenticationConfig() {
+  }
+
+  public static X509AuthenticationConfig getInstance() {
+    if (instance == null) {
+      synchronized (X509AuthenticationConfig.class) {
+        if (instance == null) {
+          instance = new X509AuthenticationConfig();
+        }
+      }
+    }
+    return instance;
+  }
+
+  /**
+   * The following System Property keys are used to extract clientId from the client cert.
+   */
+  public static final String SSL_X509_CONFIG_PREFIX = "zookeeper.ssl.x509.";
+  public static final String SSL_X509_CLIENT_CERT_ID_TYPE =
+      SSL_X509_CONFIG_PREFIX + "clientCertIdType";
+  public static final String SSL_X509_CLIENT_CERT_ID_SAN_MATCH_TYPE =
+      SSL_X509_CONFIG_PREFIX + "clientCertIdSanMatchType";
+  // Match Regex is used to choose which entry to use, default value ".*"
+  public static final String SSL_X509_CLIENT_CERT_ID_SAN_MATCH_REGEX =
+      SSL_X509_CONFIG_PREFIX + "clientCertIdSanMatchRegex";
+  // Extract Regex is used to construct a client ID (acl entity) to return, default value ".*"
+  public static final String SSL_X509_CLIENT_CERT_ID_SAN_EXTRACT_REGEX =
+      SSL_X509_CONFIG_PREFIX + "clientCertIdSanExtractRegex";
+  // Specifies match group index for the extract regex (i in Matcher.group(i)), default value 0
+  public static final String SSL_X509_CLIENT_CERT_ID_SAN_EXTRACT_MATCHER_GROUP_INDEX =
+      SSL_X509_CONFIG_PREFIX + "clientCertIdSanExtractMatcherGroupIndex";
+  public static final String SUBJECT_ALTERNATIVE_NAME_SHORT = "SAN";
+  private static final String DEFAULT_REGEX = ".*";
+  private String clientCertIdType;
+  private int clientCertIdSanMatchType = -1;
+  private String clientCertIdSanMatchRegex;
+  private String clientCertIdSanExtractRegex;
+  private int clientCertIdSanExtractMatcherGroupIndex = -1;
+
+  /** ZooKeeper server-side ZNode group ACL feature */
+//   x509ClientIdAsAclEnabled enables/disables whether znodes created by auth'ed clients
+//   should have ACL fields populated with the client Id given by the authentication provider.
+//   Has the same effect as the ZK client using ZooDefs.Ids.CREATOR_ALL_ACL.
+  public static final String ZNODE_GROUP_ACL_CONFIG_PREFIX = "zookeeper.X509ZNodeGroupAclProvider.";
+  // Enables/disables whether znodes created by auth'ed clients
+  // should have ACL fields populated with the client Id given by the authentication provider.
+  // Has the same effect as the ZK client using ZooDefs.Ids.CREATOR_ALL_ACL.
+  public static final String SET_X509_CLIENT_ID_AS_ACL =
+      ZNODE_GROUP_ACL_CONFIG_PREFIX + "setX509ClientIdAsAcl";
+  // A list of domain names that will have cross-domain access privilege, separated by ","
+  public static final String SUPER_USER_DOMAIN_NAME =
```

Comment:
I think we discussed how "super user domain" is a confusing term - especially since we have a notion of "super user". What do you think about renaming it to something like "cross-domain/multi-domain"?